providers/ms_entra_id.go

```diff
@@ -0,0 +1,162 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/util"
+	"github.com/spf13/cast"
+)
+
+// MicrosoftEntraIDProvider represents provider for Azure Entra Authentication V2 endpoint
+type MicrosoftEntraIDProvider struct {
+	*OIDCProvider
+	skipGraphGroups           bool
+	multiTenantAllowedTenants []string
+
+	microsoftGraphURL *url.URL
+}
+
+const (
+	microsoftEntraIDProviderName = "Microsoft Entra ID"
+)
+
+var (
+	microsoftGraphURL = &url.URL{
+		Scheme: "https",
+		Host:   "graph.microsoft.com",
+		Path:   "/v1.0/me",
+	}
+)
+
+// NewMicrosoftEntraIDProvider initiates a new MicrosoftEntraIDProvider
+func NewMicrosoftEntraIDProvider(p *ProviderData, opts options.Provider) *MicrosoftEntraIDProvider {
+	p.setProviderDefaults(providerDefaults{
+		name: microsoftEntraIDProviderName,
+	})
+
+	return &MicrosoftEntraIDProvider{
+		OIDCProvider: NewOIDCProvider(p, opts.OIDCConfig),
+
+		skipGraphGroups:           opts.MicrosoftEntraIDConfig.DisableGroupsFromGraph,
+		multiTenantAllowedTenants: opts.MicrosoftEntraIDConfig.AllowedMultiTenants,
+		microsoftGraphURL:         microsoftGraphURL,
+	}
+}
+
+// EnrichSession checks for group overage and passes through to generic EnrichSession()
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	err := p.OIDCProvider.EnrichSession(ctx, session)
+
```

Comment:
Regarding issuer verification:

`EnrichSession` is always called no matter if it's the first authentication or following reqeuests.

Therefore we can add a verification in here to check against every tenant id provided by the config. We will need to parse the JWT token in `session.IDToken` can be as simple as:

```go
	rawPayload, _ := base64.StdEncoding.DecodeString(strings.Split(session.IDToken, ".")[1])

	payload := make(map[string]interface{})
	json.Unmarshal([]byte(rawPayload), &payload)

	issuer := payload["iss"].(string)
```

https://go.dev/play/p/TAi9QePsJ7_L

Reply:
The concern I've mentioned on the call is [OpenID Provider Configuration Validation](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationValidation)  that (I believe) happens here :fist_right: [provider.go#L63
](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/providers/oidc/provider.go#L63) and happens on the provider initialization. It's disabled by the same flag that skips validation of the ID token which is a completely separate process. I will include the detailed information on the documentation.

Currently check against the Issuer URL is implemented in `ValidateSession()`, you think this should be moved to `EnrichSession()`?

Reply:
You are right and for multi tenancy we have to disable odic verification because in that case the common tenant url is used but the resulting oidc token will contain a specific issuer which we can then check inside the enrich session method. it will act as a second layer of verification.

Reply:
and indeed i got so into the generic implementation that i totally missed that you already implemented my idea using the proper `ValidateSession` flow.

Reply:
I'll go through everything once more tomorrow but I think we are ready to merge 🎉